### PR TITLE
Add `leadingNames` and `acceptingAnyToken` properties to `Parser` for position-aware collision detection

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

This PR adds two new properties to the `Parser` interface so that `runParser()` can detect meta-name collisions (help, version, completion) using structural parser semantics instead of inferring token positions from the display-oriented `usage` tree.

The previous approach extracted leading names by traversing the `usage` tree with `extractLeadingOptionNames()`, `extractLeadingCommandNames()`, and `extractLeadingLiteralValues()`. This was unreliable because `usage` is a display-oriented structure: `tuple()` sorts it by priority, `command()` spreads inner parser usage as siblings, and `conditional()` may or may not produce `literal` terms depending on the discriminator type. These representational choices caused both false positives and false negatives in collision detection.

The new approach gives each combinator a `leadingNames: ReadonlySet<string>` property that reports which names it could match at `argv[0]`, computed from its own parsing semantics. A companion `acceptingAnyToken: boolean` property indicates whether the parser unconditionally consumes any positional token (like `argument()` does), which lets shared-buffer compositions (`tuple()`, `object()`, `merge()`, `concat()`) exclude names that are unreachable behind a catch-all parser.

## Design

**`leadingNames`** contains the option names, command names, and literal values that this parser could match at the first buffer position. Each combinator computes it structurally:

- Primitives report their own names: `option("-v", "--verbose", string())` produces `{"-v", "--verbose"}`, `command("help", ...)` produces `{"help"}`, `argument()` produces `{}` (no specific name).
- Alternatives (`or()`, `longestMatch()`) union all branches.
- Shared-buffer compositions (`tuple()`, `object()`, `merge()`, `concat()`) process children in priority order and stop adding positional names after encountering a catch-all parser.
- Wrappers (`optional()`, `map()`, `group()`, etc.) forward from the inner parser.
- `conditional()` includes the discriminator's names plus the default branch's names (since the default branch receives the original buffer when the discriminator fails or its value has no matching branch key).
- `command("tool", innerParser)` reports only `{"tool"}`, not the inner parser's names.

**`acceptingAnyToken`** is `true` only for parsers that consume any positional token regardless of content (primarily `argument()`). Since `argument()` rejects option-like tokens (those starting with `"-"`), this flag acts as a positional catch-all, not a universal one. In `sharedBufferLeadingNames()`, encountering a catch-all parser blocks subsequent positional names but still allows option-like names through. Wrappers that can succeed without consuming (`optional()`, `withDefault()`, `multiple()` with `min = 0`) always set this to `false`.

**`UserParserNames`** is simplified from six fields (`leadingOptions`, `leadingCommands`, `leadingLiterals`, `allOptions`, `allCommands`, `allLiterals`) to four (`leadingNames`, `allOptions`, `allCommands`, `allLiterals`). For command-form meta entries, `validateMetaNameCollisions()` checks the flat `leadingNames` set and classifies each colliding name by cross-referencing with `allOptions`/`allCommands` for the error message.

## Example

```typescript
// Before: tuple() sorted usage by priority, placing command("help")
// before argument() in the usage tree. extractLeadingCommandNames()
// then reported "help" as a leading command even though argument()
// would consume it first at runtime.
const parser = tuple([
  or(argument(string()), command("foo", object({}))),
  command("help", object({})),
]);

// After: leadingNames = {"foo"} (not {"foo", "help"}).
// or(argument(), command("foo")) has acceptingAnyToken = true,
// so command("help") at the same priority is blocked for positional
// names. This parser no longer falsely collides with a built-in
// help command.
```

## Breaking changes

- `Parser` interface now requires `leadingNames` and `acceptingAnyToken` properties. Custom parser implementations must include these. The docs example in *docs/concepts/extend.md* has been updated accordingly.
- `extractLeadingOptionNames()`, `extractLeadingCommandNames()`, and `extractLeadingLiteralValues()` are removed from `@optique/core/usage`. These were added in the unreleased 1.0.0 cycle and have no downstream consumers. Use `Parser.leadingNames` instead.
- `UserParserNames` interface is simplified: `leadingOptions`, `leadingCommands`, and `leadingLiterals` are replaced by a single `leadingNames` field.

## Test plan

- [x] `mise test:deno` (394 tests pass)
- [x] `mise run check` (type check, lint, format, dry-run publish)
- [x] `pnpm --dir docs build` (Twoslash type checking in VitePress docs)

Closes https://github.com/dahlia/optique/issues/735

### CODE REVIEW

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.ts [hunk 1]
```diff
@@ -8122,6 +8193,15 @@ export function conditional(
...
     usage,
+    // The default branch receives the original buffer both when the
+    // discriminator fails outright AND when it succeeds but its value
+    // does not match any branch key.  In either case the default
+    // branch's leading names are reachable at position 0.
+    leadingNames: defaultBranch
+      ? unionLeadingNames([discriminator, defaultBranch])
+      : discriminator.leadingNames,
+    acceptingAnyToken: discriminator.acceptingAnyToken ||
+      (defaultBranch?.acceptingAnyToken ?? false),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in 09785649. `conditional()` can return `success: false, consumed: 0` after the discriminator consumes (e.g., when the discriminator value has no matching branch key and the default branch rejects). This means it is never a reliable catch-all, so `acceptingAnyToken` is now always `false`.

File: docs/concepts/extend.md [hunk 2]
```diff
@@ -198,6 +200,24 @@ which includes any annotations passed to `parse()`. By calling
...
+    the parser matches as its first token.  If the parser accepts
+    *any* token (like `argument()`), use an empty set and set
+    `acceptingAnyToken` instead.  Built-in combinators compute this
+    automatically.
+
+`acceptingAnyToken`
+:   Set to `true` when the parser unconditionally consumes the first
+    token regardless of its content.  This tells shared-buffer
+    compositions (`tuple()`, `object()`, etc.) that sibling parsers
+    with equal or lower priority cannot match at the same position.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/extend.md]: Fixed in 09785649. Updated the docs to say "any positional token" and "positional token regardless of its value" to clarify that `acceptingAnyToken` does not cover option-like tokens.

File: packages/core/src/validate.ts [hunk 3]
```diff
@@ -249,78 +244,89 @@ export function validateMetaNameCollisions(
...
+            `User-defined command "${name}" conflicts with the ` +
+              `built-in ${label}.`,
+          );
+        }
+        if (userNames.allLiterals.has(name)) {
+          throw new TypeError(
+            `Literal value "${name}" conflicts with the ` +
+              `built-in ${label}.`,
+          );
+        }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/validate.ts:299]: The option-form branch (lines 277-294) already uses separate `if` blocks that each throw independently, which is the guard clause pattern. The command-form branch (lines 256-273) uses `if`/`else if`/`else` intentionally because it classifies a single name from `leadingNames` into one of three mutually exclusive categories for the error message. Converting this to separate `if` blocks would not improve clarity since at most one classification applies per name.

File: packages/core/src/constructs.ts [hunk 4]
```diff
@@ -61,6 +61,63 @@ type CombineObjectModes<
...
+        // option-like names remain reachable because argument() rejects
+        // tokens that start with "-".
+        if (positionalBlocked && !name.startsWith("-")) continue;
+        names.add(name);
+      }
+    }
+    if (p.acceptingAnyToken) positionalBlocked = true;
+  }
+  return names.size === 0 ? EMPTY_LEADING_NAMES : names;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:119]: Since `leadingNames` and `acceptingAnyToken` are required members of the `Parser` interface, TypeScript already rejects custom parsers that omit them at compile time. Adding runtime normalization would be redundant and would mask type errors that the compiler should catch. The `LeadingNameSource` interface mirrors the `Parser` shape and is only used internally with parsers that have already passed type checking.

File: packages/core/src/facade.test.ts [hunk 5]
```diff
@@ -5388,6 +5388,8 @@ describe("runWith", () => {
         $stateType: [] as readonly (string | undefined)[],
         priority: 1,
         usage: [],
+        leadingNames: new Set(),
+        acceptingAnyToken: false,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts:5392]: `new Set()` here gets contextual typing from the enclosing object literal which has an explicit `Parser<...>` type annotation, so TypeScript infers `Set<string>` correctly. The type checker passes without errors. Adding explicit `<string>` would not change the compiled output or type safety.

File: packages/core/src/modifiers.ts [hunk 6]
```diff
@@ -239,6 +239,10 @@ export function optional<M extends Mode, TValue, TState>(
     placeholder: undefined as TValue | undefined,
     priority: parser.priority,
     usage: [{ type: "optional", terms: parser.usage }],
+    leadingNames: parser.leadingNames,
+    // optional/withDefault can succeed without consuming, so the inner
+    // parser's catch-all status does not apply to the wrapper.
+    acceptingAnyToken: false,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.ts:245]: `nonEmpty()` takes `Parser<M, readonly T[], TState>` (array result from `multiple()`), so `nonEmpty(optional(argument()))` does not type-check since `optional(argument())` returns `string | undefined`, not an array. The concrete example `concat(nonEmpty(optional(argument(string()))), ...)` also fails at the type level.

In practice, `nonEmpty` wraps `multiple`: `nonEmpty(multiple(argument()))`. Here `multiple(min=0)` already sets `acceptingAnyToken: false`, and `nonEmpty` propagates it. This means the result is not marked as a catch-all. The false positive (including sibling names that are technically blocked) is in the safe direction and only matters when parsers have equal priority, which is uncommon for argument-based compositions since `argument()` has priority 5 while commands and options have 10-15.

**Comment by dahlia** [comment]
@codex review

File: docs/concepts/extend.md [hunk 7]
```diff
@@ -198,6 +200,25 @@ which includes any annotations passed to `parse()`. By calling
...
+    this automatically.
+
+`acceptingAnyToken`
+:   Set to `true` when the parser unconditionally consumes the first
+    positional token regardless of its value.  This tells shared-buffer
+    compositions (`tuple()`, `object()`, etc.) that sibling parsers
+    with equal or lower priority cannot match at the same position
+    for positional (non-option) tokens.
+    Most custom parsers should set this to `false`.
+
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/extend.md]: Fixed in ee9a3bab. Reworded the intro to clarify that `leadingNames` enables collision detection while `acceptingAnyToken` controls composition semantics.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/modifiers.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.ts:245]: The described scenario is not a false positive. In `tuple([optional(or(argument(), command("foo"))), command("help", ...)])` with a built-in help command enabled, `runParser()` installs a meta help parser that intercepts `"help"` at `argv[0]` *before* the user parser runs. So the user's `command("help")` is genuinely unreachable when the input starts with `"help"`, and flagging the collision is correct.

The `optional()` wrapper correctly sets `acceptingAnyToken: false` because it can succeed without consuming. The fact that its inner parser would consume positional tokens when present does not change the meta-level collision: the built-in help command always wins at `argv[0]`.

File: packages/core/src/constructs.test.ts [hunk 8]
```diff
@@ -9222,6 +9252,7 @@ describe("branch coverage: constructs.ts edge cases", () => {
         $stateType: undefined,
         priority: 0,
         usage: [],
+        leadingNames: new Set(),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:9255]: Fixed in e60b2a1d. Added `acceptingAnyToken: false` to both mock parsers.

File: packages/core/src/constructs.test.ts [hunk 9]
```diff
@@ -9313,6 +9344,7 @@ describe("branch coverage: constructs.ts edge cases", () => {
...
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:9348]: Fixed in the same commit (e60b2a1d).

**Comment by dahlia** [comment]
@codex review

**Comment by dahlia** [comment]
@coderabbitai full review

File: CHANGES.md [hunk 10]
```diff
@@ -115,32 +115,40 @@ To be released.
...
+    tree.  Shared-buffer compositions (`tuple()`, `object()`, `merge()`,
+    `concat()`) use priority ordering and `acceptingAnyToken` to exclude
+    names that are unreachable behind a catch-all parser like `argument()`.
+    `conditional()` with a default branch now includes the default branch's
+    leading names.  This replaces the usage-tree-based
+    `extractLeadingOptionNames()`, `extractLeadingCommandNames()`, and
+    `extractLeadingLiteralValues()`, which produced incorrect results for
+    `tuple()` (priority-based sorting), `command()` (inner usage
+    spreading), and `conditional()` (missing literal terms for argument
+    discriminators).  [[#735], [#741]]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md:138]: Fixed in 33bccd31. Updated the CHANGES.md entry to say "leading tokens (option names, command names, and literal values)".

File: packages/core/src/constructs.test.ts [hunk 11]
```diff
@@ -11366,3 +11444,194 @@ describe("merge()/concat() suggest with cross-parser dependencies", () => {
...
+    );
+    assert.deepEqual(parser.leadingNames, new Set(["--verbose", "help"]));
+  });
+
+  it("should be the union of all parsers for concat()", () => {
+    const parser = concat(
+      tuple([flag("--verbose")]),
+      tuple([command("help", object({}))]),
+    );
+    assert.deepEqual(parser.leadingNames, new Set(["--verbose", "help"]));
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:11529]: Added in 33bccd31. Both `merge()` and `concat()` now have catch-all regression tests that verify positional names are blocked while option-like names pass through.

File: packages/core/src/constructs.ts [hunk 12]
```diff
@@ -8122,6 +8193,17 @@ export function conditional(
...
+    // discriminator fails outright AND when it succeeds but its value
+    // does not match any branch key.  In either case the default
+    // branch's leading names are reachable at position 0.
+    leadingNames: defaultBranch
+      ? unionLeadingNames([discriminator, defaultBranch])
+      : discriminator.leadingNames,
+    // conditional() can fail without consuming (e.g., discriminator
+    // consumes but its value has no matching branch and the default
+    // branch rejects), so it is never a reliable catch-all.
+    acceptingAnyToken: false,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in 33bccd31. `conditional().acceptingAnyToken` now derives from `defaultBranch?.acceptingAnyToken ?? false`. A catch-all default branch makes the conditional consume any positional token because it reparses the original buffer. The discriminator's own catch-all status does not matter since it only routes to branches.

File: packages/core/src/validate.ts [hunk 13]
```diff
@@ -249,78 +244,89 @@ export function validateMetaNameCollisions(
...
-        for (const literal of literalNames) {
-          if (literal.startsWith(prefix)) {
+          } else {
             throw new TypeError(
-              `Literal value "${literal}" conflicts with the ` +
-                `built-in ${label} (prefix "${prefix}").`,
+              `User-defined name "${name}" conflicts with the ` +
+                `built-in ${label}.`,
             );
           }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/validate.ts:277]: Fixed in c9bd0cc2. Added `allLiterals` check between the `allCommands` and fallback branches so that literal values get a "Literal value" error message instead of the generic "User-defined name".

File: packages/core/src/constructs.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in 33bccd31 (prior push). `conditional().acceptingAnyToken` now derives from `defaultBranch?.acceptingAnyToken ?? false` instead of being hard-coded to `false`.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.test.ts [hunk 14]
```diff
@@ -11366,3 +11444,216 @@ describe("merge()/concat() suggest with cross-parser dependencies", () => {
...
+    const parser = command(
+      "tool",
+      longestMatch(
+        command("help", object({})),
+        command("build", object({})),
+      ),
+    );
+    assert.deepEqual(parser.leadingNames, new Set(["tool"]));
+  });
+});
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:11600]: `argument(choice(["server", "client"]))` has `leadingNames = new Set()` because `argument()` accepts any positional token, not specifically "server" or "client". The `choice()` constraint is a value-level validation that happens after the token is consumed, not a structural name declaration. Branch keys ("server", "client") are runtime discriminator values, not structural parser names. Detecting branch key collisions requires a separate mechanism tracked in [#734](https://github.com/dahlia/optique/issues/734).

File: packages/core/src/modifiers.ts [hunk 15]
```diff
@@ -1666,6 +1678,8 @@ export function nonEmpty<M extends Mode, T, TState>(
     $stateType: parser.$stateType,
     priority: parser.priority,
     usage: parser.usage,
+    leadingNames: parser.leadingNames,
+    acceptingAnyToken: parser.acceptingAnyToken,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/modifiers.ts:1682]: This is a broader refactoring outside the scope of this PR. The explicit property forwarding is intentional: wrappers like `optional()` and `withDefault()` must set `acceptingAnyToken: false` explicitly (not inherited from the inner parser), and `multiple()` computes it conditionally based on `min`. Using spread would risk silently propagating values that should be overridden, requiring careful exclusion patterns that are harder to audit than explicit assignment.
